packages/experiments/src/components/Button/MenuButton/MenuButton.state.ts

```diff
@@ -1,68 +1,56 @@
-import * as React from 'react';
-import { BaseState } from '../../../utilities/BaseState';
-import { IMenuButtonProps, IMenuButtonViewProps } from './MenuButton.types';
-
-export type IMenuButtonState = Pick<
-  IMenuButtonViewProps,
-  'expanded' | 'onClick' | 'onKeyDown' | 'buttonRef' | 'onMenuDismiss' | 'menuTarget'
->;
-
-export class MenuButtonState extends BaseState<IMenuButtonProps, IMenuButtonViewProps, IMenuButtonState> {
-  private _buttonRef = React.createRef<HTMLButtonElement>();
-
-  constructor(props: MenuButtonState['props']) {
-    super(props, {
-      controlledProps: ['expanded']
-    });
-
-    this.state = {
-      expanded: !!props.defaultExpanded,
-      onClick: this._onClick,
-      onKeyDown: this._onKeyDown,
-      buttonRef: this._buttonRef,
-      onMenuDismiss: this._onMenuDismiss,
-      menuTarget: undefined
-    };
-  }
-
-  public focus(): void {
-    if (this._buttonRef.current) {
-      this._buttonRef.current.focus();
-    }
-  }
-
-  private _onMenuDismiss = () => {
-    this.setState({
-      expanded: false
-    });
-  };
-
-  private _onClick = (ev: React.MouseEvent<HTMLElement>) => {
-    const { disabled, onClick } = this.props;
-
-    if (!disabled) {
-      if (onClick) {
-        onClick(ev);
-
-        if (ev.defaultPrevented) {
-          return;
+import { useCallback, useState } from 'react';
+import { useControlledState } from '../../../Foundation';
+import { KeyCodes } from '../../../Utilities';
+import { IMenuButtonComponent, IMenuButtonViewProps } from './MenuButton.types';
+
+export const useMenuButtonState: IMenuButtonComponent['state'] = props => {
+  const [menuTarget, setMenuTarget] = useState<HTMLElement | undefined>(undefined);
+  const [expanded, setExpanded] = useControlledState(props, 'expanded', {
+    defaultPropName: 'defaultExpanded',
+    defaultPropValue: false
+  });
+
+  const _onMenuDismiss = useCallback(() => {
+    setExpanded(false);
+  }, []);
+
+  const { disabled, onClick } = props;
+
+  const _onClick = useCallback(
```

Comment:
One thing I wonder, would it be worthwhile to put this into their own state hook as well? Mainly to help set the precedent that we want generic helper functions that can act as building blocks.

Reply:
I'd thought about it, but it seems behavior is so varied that the helpers would be so basic (or the options so complex) that I wonder if there is a net positive in having them. For example, do you always call callbacks? Do you call callbacks only if not processed? Do you stop propagation? Etc.

Reply:
Like this would be a ` useExpandMenu ` hook that could be swapped around whenever we want a menu to get expanded.

Reply:
I just feel like if we want there to be lots of shareable code between components, we should start trying now to set the standard that we want things to be in a common area. An example of where I would want this is for combobox, it should behave similarly IMO.

Reply:
I agree and I've done that where I think it makes sense, like `useControlledState`. Here, though, we'd have to pass `disabled`, `onClick`, `expanded`, `setExpanded`, and `setMenuTarget` all to a helper. I 100% believe in making things common where it makes sense, but I don't think it's quite so black and white in this case. There is a balance between reuse and simplicity, and making this common doesn't add much value, I don't think. The logic is both highly customized to this component's use case while also relatively simple. Most of the implementation is actually the interacting with the same surface area (the inputs I mentioned above) which isn't meaningfully reduced by extracting it out as a separate function.